pithos/GstPlayer.py

```diff
@@ -0,0 +1,443 @@
+
+        # Reset the timer_id only if called from the timeout
+        # to avoid GLib.source_remove warnings.
+        if from_timeout:
+            self._buffering_timer_id = 0
+        buffering = self._query_buffer()
+
+        if buffering and self._actual_state is not PseudoGst.BUFFERING:
+            logging.debug('Buffer underrun')
+            if self._set_player_state(PseudoGst.BUFFERING):
+                logging.debug('Pausing pipeline')
+        elif not buffering and self._actual_state is PseudoGst.BUFFERING:
+            logging.debug('Buffer overrun')
+            if self._desired_state is PseudoGst.STOPPED:
+                if self._set_player_state(PseudoGst.PLAYING, change_gst_state=True):
+                    logging.debug('Song starting')
+            elif self._desired_state is PseudoGst.PLAYING:
+                if self._set_player_state(PseudoGst.PLAYING, change_gst_state=True):
+                    logging.debug('Restarting pipeline')
+            elif self._desired_state is PseudoGst.PAUSED:
+                if self._set_player_state(PseudoGst.PAUSED, change_gst_state=True):
+                    logging.debug('User paused')
+            # Tell everyone to update their clocks after we're done buffering or
+            # in case it takes a while after the song-changed signal for actual playback to begin.
+            self.emit('buffering-finished', self.query_position())
+        return buffering
+
+    def _set_player_state(self, target, change_gst_state=False):
+        change_gst_state = change_gst_state or self._actual_state is not PseudoGst.BUFFERING
+        if change_gst_state:
+            ret = self._player.set_state(target.state)
+            if ret == Gst.StateChangeReturn.FAILURE:
+                current_state = self._player.state_get_name(self._actual_state.state)
+                target_state = self._player.state_get_name(target.state)
+                logging.warning('Error changing player state from: {} to: {}'.format(current_state, target_state))
+                return False
+            self._actual_state = target
+        if target is not PseudoGst.BUFFERING:
+            self._desired_state = target
+        self.emit('new-player-state')
+        return True
+
+    def _on_gst_source(self, player, params):
+        soup = player.props.source.props
+        props = {
+            'iradio_mode': False,
+            'user_agent': 'io.github.Pithos',
+            'method': 'GET',
+            'compress': True,
+            'keep_alive': True,
+            'proxy': None,
+            'proxy_id': None,
+            'proxy_pw': None,
+        }
+
+        proxy = self._settings['proxy'] or urllib.request.getproxies().get('http')
+        if proxy:
+            scheme, props['proxy_id'], props['proxy_pw'], props['proxy'] = parse_proxy(proxy)
+
+        for prop, value, in props.items():
+            if value is not None and hasattr(soup, prop):
+                setattr(soup, prop, value)
+
```

Comment:
__[Flake8]__

blank line at end of file


<!-- Comment by [SideCI](https://sideci.com) -->